### ISSUE TITLE
Nemotron-3-Nano-30B-A3B: ~8K tps/gpu vs ~12K for Qwen3-30B-A3B — TE optimization for Mamba2 layers?

### ISSUE THREAD

**Issue by bhaktatejas922** [NVIDIA-NeMo/Automodel#1187]
## Summary

Training Nemotron-3-Nano-30B-A3B with TE backend on 4x B200 GPUs, getting ~8K tokens/sec/gpu. The Qwen3-30B-A3B benchmark config (`qwen3_moe_30b_te_deepep.yaml`) achieves ~12K tps/gpu on 8x GPUs with a similar setup. Wondering if there are recommended optimizations for the Nemotron hybrid architecture, or if the gap is expected.

## Setup

- **Model**: `nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-Base-BF16` (hybrid Mamba2 + MoE + Attention)
- **Hardware**: 4x NVIDIA B200 (178GB each), single node, NVLink
- **Framework**: nemo_automodel 0.3.0rc0
- **Precision**: BF16 (FP8 disabled — only 25/6005 layers were eligible with torchao)

### Config
```yaml
model:
  _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_pretrained
  pretrained_model_name_or_path: nvidia/NVIDIA-Nemotron-3-Nano-30B-A3B-Base-BF16
  torch_dtype: bf16
  trust_remote_code: true
  backend:
    _target_: nemo_automodel.components.models.common.BackendConfig
    attn: te
    linear: te
    rms_norm: te
    rope_fusion: true
    enable_deepep: false
    enable_hf_state_dict_adapter: true
    enable_fsdp_optimizations: true

distributed:
  _target_: nemo_automodel.components.distributed.fsdp2.FSDP2Manager
  dp_replicate_size: 1
  tp_size: 1
  cp_size: 1
  ep_size: 4
  sequence_parallel: false

parallelizer:
  _target_: nemo_automodel.components.moe.parallelizer.parallelize_model
  activation_checkpointing: true

compile:
  enabled: true
  mode: "default"

fp8:
  enabled: false

dataset:
  seq_length: 65536

step_scheduler:
  global_batch_size: 4
  local_batch_size: 1
```

### Results
- **~8K tps/gpu** (steady state after compile warmup)
- **130 GB/GPU** memory usage
- Loss converging normally

## Question

The Nemotron-3-Nano-30B-A3B is a hybrid architecture (Mamba2 + MoE + Attention), while Qwen3-30B-A3B is pure transformer MoE. Our understanding is that TE fused kernels (flash attention, fused linear, fused RMSNorm) accelerate the transformer layers but don't apply to Mamba2 layers (causal_conv1d, selective_scan). Similarly, torch.compile hits graph breaks on Mamba2 ops.

1. **Is the ~8K vs ~12K gap expected** due to the Mamba2 layers not benefiting from TE optimizations?
2. **Are there any recommended optimizations** for the Nemotron hybrid architecture that we might be missing? (e.g., Mamba2-specific kernel fusions, different parallelism strategies)
3. **Would native TE FP8** (via `te.autocast()`) convert more layers than torchao for this hybrid architecture? With torchao only 25/6005 layers were eligible.
4. **Does `fake_balanced_gate: true`** help for Nemotron's MoE layers specifically?

Any guidance on getting closer to the pure MoE throughput would be appreciated.

## Environment
- PyTorch 2.7.0+cu128
- TransformerEngine (latest via pip)
- CUDA 12.8
- 4x NVIDIA B200 (SM 10.0, Blackwell)
